pkg/cmd/util/tokencmd/negotiate_gssapi.go

```diff
@@ -0,0 +1,211 @@
+	}
+
+	lib, err := c.loadLib()
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Get incoming token
+	_, incomingToken, err := getNegotiateToken(headers)
+	if err != nil {
+		return nil, false, err
+	}
+	incomingTokenBuffer, err := lib.MakeBufferBytes(incomingToken)
+	if err != nil {
+		return nil, false, err
+	}
+	defer incomingTokenBuffer.Release()
+
+	if c.ctx == nil {
+		u, err := url.Parse(requestURL)
+		if err != nil {
+			return nil, false, err
+		}
+
+		glog.V(5).Infof("importing service name %s", u.Host)
+		nameBuf, err := lib.MakeBufferString("HTTP@" + u.Host)
+		if err != nil {
+			return nil, false, err
+		}
+		defer nameBuf.Release()
+
+		name, err := nameBuf.Name(lib.GSS_C_NT_HOSTBASED_SERVICE)
+		if err != nil {
+			return nil, false, err
+		}
+		c.name = name
+		c.ctx = lib.GSS_C_NO_CONTEXT
+	}
+
+	// Call InitSecContext to validate the incoming token (if given), and get our outgoing token (if needed)
+	// Save the returned context
+	// TODO: validate returned flags support requested services
+	var outgoingToken *gssapi.Buffer
+	c.ctx, _, outgoingToken, _, _, err = lib.InitSecContext(lib.GSS_C_NO_CREDENTIAL, c.ctx, c.name, lib.GSS_C_NO_OID, c.flags, time.Duration(0), lib.GSS_C_NO_CHANNEL_BINDINGS, incomingTokenBuffer)
+	defer outgoingToken.Release()
+
+	switch err {
+	case nil:
+		glog.V(5).Infof("InitSecContext returned GSS_S_COMPLETE")
+		c.continueNeeded = false
+	case gssapi.ErrContinueNeeded:
+		glog.V(5).Infof("InitSecContext returned GSS_S_CONTINUE_NEEDED")
+		c.continueNeeded = true
+	default:
+		glog.V(5).Infof("InitSecContext returned error: %v (%#v)", err, err)
+		return nil, false, err
+	}
+
+	responseHeaders := http.Header{}
+	responseHeaders.Set("Authorization", "Negotiate "+base64.StdEncoding.EncodeToString(outgoingToken.Bytes()))
+	return responseHeaders, true, nil
+}
+
+func (c *NegotiateChallengeHandler) CompleteChallenge(requestURL string, headers http.Header) error {
+	if c.released {
+		glog.V(5).Infof("already released")
+		return nil
+	}
+
+	lib, err := c.loadLib()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		glog.V(5).Infof("releasing name and context")
+		c.name.Release()
+		c.ctx.Release()
+		glog.V(5).Infof("unloading gssapi library")
+		c.lib.Unload()
+		c.released = true
+	}()
+
+	if !c.continueNeeded {
+		return nil
+	}
+
+	glog.V(5).Infof("continue needed")
+	_, incomingToken, err := getNegotiateToken(headers)
+	if err != nil {
+		return err
+	}
+	incomingTokenBuffer, err := lib.MakeBufferBytes(incomingToken)
+	if err != nil {
+		return err
+	}
+	defer incomingTokenBuffer.Release()
+
+	// Call InitSecContext to validate the incoming token (if given), and get our outgoing token (if needed)
+	// Save the returned context
+	var outgoingToken *gssapi.Buffer
+	c.ctx, _, outgoingToken, c.flags, _, err = lib.InitSecContext(lib.GSS_C_NO_CREDENTIAL, c.ctx, c.name, lib.GSS_C_NO_OID, c.flags, time.Duration(0), lib.GSS_C_NO_CHANNEL_BINDINGS, incomingTokenBuffer)
```

Comment:
Correction, I see why this is necessary. The underlying library unconditionally allocates the memory for this buffer, so we have to retrieve it in order to free it so we don't leak.